File: src/core/MRReviewer.ts
import readline from 'node:readline';

import { GitAdapterFactory } from '../adapters/git';
import type { GitAdapter } from '../adapters/git/GitAdapter';
import { LLMAdapterFactory } from '../adapters/llm';
import type { LLMAdapter } from '../adapters/llm/LLMAdapter';
import type { AppConfig } from '../config/config';
import { FileUtils } from '../utils/fileUtils';
import { Logger } from '../utils/logging';
import { ReviewFormatter } from './ReviewFormatter';
import type { FileReview, Overview, ReviewFeedback } from './ReviewFormatter';

interface ReviewMetadata {
  timestamp: string;
  llmProvider: string;
  llmModel: string;
  mrMode: string;
  gitPlatform: string;
  commandLine: string;
  [key: string]: string;
}

interface ParsedFeedback extends ReviewFeedback {
  [key: string]: any;
}

/**
 * Main class for performing MR/PR reviews
 */
export class MRReviewer {
  private config: AppConfig;
  private llmAdapter: LLMAdapter;
  private gitAdapter: GitAdapter;
  private formatter: ReviewFormatter;
  private fileUtils: FileUtils;
  private logger: Logger;
  private promptTemplate: string;

  /**
   * Create a new MR reviewer
   * @param config - The loaded and validated configuration
   */
  constructor(config: AppConfig) {
    this.config = config;
    this.logger = new Logger(config.verbose);
    this.fileUtils = new FileUtils();

    // Create adapters using factories
    const llmFactory = new LLMAdapterFactory();
    const gitFactory = new GitAdapterFactory();

    this.llmAdapter = llmFactory.create(config);
    this.gitAdapter = gitFactory.create(config);

    // Create formatter with the configured template
    this.formatter = new ReviewFormatter(config.formatterTemplate);

    // Load prompt template from directory
    this.promptTemplate = this.fileUtils.loadPromptFromDirectory(config.promptDir);
  }

  /**
   * Execute the review workflow
   * @returns Path to the saved review file
   */
  async review(): Promise<string | null> {
    try {
      // Get the diff
      const diff = await this.getDiff();
      if (this.config.verbose) {
        this.logger.debug(`</diff start>${diff}</diff end>`);
      }
      this.logger.info(`Retrieved diff (${diff.length} characters)`);

      if (diff.length === 0) {
        this.logger.warn('No changes detected, skipping LLM analysis');
        return null;
      }

      if (await this.shouldCancelReview(diff)) {
        return null;
      }

      const feedback = await this.analyzeDiff(diff);
      this.logger.info('LLM analysis completed');

      const metadata = this.createReviewMetadata();

      const reviewFilePath = this.fileUtils.saveReviewToFile(
        this.config.reviewsDir,
        feedback,
        metadata
      );

      const parsedFeedback = this.parseFeedback(feedback);
      const formattedComment = this.formatter.format({
        metadata,
        feedback: parsedFeedback,
      });

      await this.postComment(formattedComment);

      this.logger.info('Review completed successfully');
      return reviewFilePath;
    } catch (error) {
      this.logger.error('Review failed:', error as Error);
      throw error;
    }
  }

  /**
   * Create metadata for the review
   *
   * @returns Review metadata object
   */
  private createReviewMetadata(): ReviewMetadata {
    return {
      timestamp: new Date().toISOString(),
      llmProvider: this.config.llmProvider,
      llmModel: this.getLLMModelName(),
      mrMode: this.config.mrMode,
      gitPlatform: this.config.gitPlatform,
      commandLine: process.argv.join(' '),
    };
  }

  /**
   * Check if the review should be canceled based on user input
   *
   * @param diff The diff content
   * @returns True if the review should be canceled
   */
  private async shouldCancelReview(diff: string): Promise<boolean> {
    if (!this.config.showDiff) {
      return false;
    }

    this.displayDiff(diff);

    // If we're not in interactive mode, continue with the review
    if (!process.stdin.isTTY) {
      return false;
    }

    const shouldCancel = await this.promptForCancellation();
    if (shouldCancel) {
      this.logger.info('LLM analysis cancelled by user');
      return true;
    }

    return false;
  }

  /**
   * Display the diff to the user
   *
   * @param diff The diff content
   */
  private displayDiff(diff: string): void {
    this.logger.user('\n===== DIFF START =====\n');
    this.logger.user(diff);
    this.logger.user('\n===== DIFF END =====\n');
  }

  /**
   * Prompt the user for cancellation
   *
   * @returns True if the user wants to cancel
   */
  private async promptForCancellation(): Promise<boolean> {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    const answer = await new Promise<string>(resolve => {
      rl.question('Continue with LLM analysis? (Y/n): ', resolve);
    });

    rl.close();
    return answer.toLowerCase() === 'n';
  }

  /**
   * Parse the feedback string into a structured object
   *
   * @param feedback The feedback string from the LLM
   * @returns Parsed feedback object
   */
  private parseFeedback(feedback: string): ParsedFeedback {
    try {
      const parsedFeedback: any = typeof feedback === 'string' ? JSON.parse(feedback) : feedback;

      // TODO(FIXME)
      // Create a new object with default values and spread the parsed feedback
      // This avoids the "specified more than once" TypeScript warning
      const result: ParsedFeedback = {
        // Default values
        overview: parsedFeedback.overview || ({} as Overview),
        fileReviews: parsedFeedback.fileReviews || ([] as FileReview[]),
        testReview: parsedFeedback.testReview || '',
        generalFeedback: parsedFeedback.generalFeedback || '',

        // Copy any other properties
        ...Object.fromEntries(
          Object.entries(parsedFeedback).filter(
            ([key]) => !['overview', 'fileReviews', 'testReview', 'generalFeedback'].includes(key)
          )
        ),
      };

      return result;
    } catch (error) {
      this.logger.error('Error parsing feedback for formatting:', error as Error);
      throw new Error('Could not format review due to JSON parsing error');
    }
  }

  /**
   * Get MR/PR diff based on the configured mode
   * @returns The diff content
   */
  async getDiff(): Promise<string> {
    try {
      if (this.isRunningInCI()) {
        return await this.getCIDiff();
      }

      if (this.config.mrMode === 'url') {
        return await this.getURLDiff();
      }

      // Default to local mode
      this.logger.info('Running in local mode, fetching diff from local git');
      return await this.gitAdapter.getLocalDiff();
    } catch (error) {
      this.logger.error('Error getting MR/PR diff:', error as Error);
      throw error;
    }
  }

  /**
   * Check if running in CI mode
   *
   * @returns True if running in CI mode
   */
  private isRunningInCI(): boolean {
    return (
      this.config.mrMode === 'ci' && process.env['CI_PIPELINE_SOURCE'] === 'merge_request_event'
    );
  }

  /**
   * Get diff for CI mode
   *
   * @returns The diff content
   */
  private async getCIDiff(): Promise<string> {
    this.logger.info('Running in CI mode, fetching diff from Git platform API');

    if (this.config.gitPlatform === 'gitlab') {
      return await this.gitAdapter.getRequestDiff({
        projectId: this.config.projectId || '',
        mergeRequestIid: this.config.mergeRequestIid || '',
      });
    }

    if (this.config.gitPlatform === 'github') {
      throw new Error('GitHub CI mode not fully implemented yet');
    }

    throw new Error('Unsupported Git platform for CI mode');
  }

  /**
   * Get diff for URL mode
   *
   * @returns The diff content
   */
  private async getURLDiff(): Promise<string> {
    if (!this.config.gitMrUrl) {
      throw new Error('Git MR/PR URL is required for URL mode');
    }

    const url = this.config.gitMrUrl.toLowerCase();
    const actualPlatform = this.config.gitPlatform;

    if (this.isGitHubUrl(url)) {
      console.log('isGitHub OK');
      process.exit(1);
      return await this.getGitHubDiffWithTempAdapter();
    }

    if (this.isGitLabUrl(url)) {
      console.log('isGitLabUrl OK');
      return await this.getGitLabDiffWithTempAdapter();
    }
    console.log('how?');
    process.exit(1);
    this.logger.info(`Running in URL mode, fetching diff from ${actualPlatform} API using URL`);

    // Use the default adapter
    const parsedUrl = this.gitAdapter.parseRequestUrl(this.config.gitMrUrl ?? '');
    return await this.gitAdapter.getRequestDiff(parsedUrl);
  }

  /**
   * Get GitHub diff using a temporary adapter
   *
   * @returns The diff content
   */
  private async getGitHubDiffWithTempAdapter(): Promise<string> {
    this.logger.info(
      'URL appears to be GitHub, but platform is set to GitLab. Switching to GitHub adapter.'
    );

    if (!this.config.gitMrUrl) {
      throw new Error('Git MR/PR URL is required for GitHub adapter');
    }

    const gitFactory = new GitAdapterFactory();
    const tempConfig = { ...this.config, gitPlatform: 'github' as const };
    const githubAdapter = gitFactory.create(tempConfig);

    const parsedUrl = githubAdapter.parseRequestUrl(this.config.gitMrUrl);
    return await githubAdapter.getRequestDiff(parsedUrl);
  }

  /**
   * Get GitLab diff using a temporary adapter
   *
   * @returns The diff content
   */
  private async getGitLabDiffWithTempAdapter(): Promise<string> {
    this.logger.info(
      'URL appears to be GitLab, but platform is set to GitHub. Switching to GitLab adapter.'
    );

    if (!this.config.gitMrUrl) {
      throw new Error('Git MR/PR URL is required for GitLab adapter');
    }

    const gitFactory = new GitAdapterFactory();
    const tempConfig = { ...this.config, gitPlatform: 'gitlab' as const };
    const gitlabAdapter = gitFactory.create(tempConfig);
    const parsedUrl = gitlabAdapter.parseRequestUrl(this.config.gitMrUrl);
    const diffResult = await gitlabAdapter.getRequestDiff(parsedUrl);
    if (this.config.verbose) {
      this.logger.debug(`</diff start>${diffResult}</diff end>`);
    }
    return diffResult;
  }

  /**
   * Check if URL is a GitHub URL
   *
   * @param url URL to check
   * @returns True if URL is a GitHub URL
   */
  private isGitHubUrl(url: string): boolean {
    return url.includes('github.com') || url.includes('/pull/');
  }

  /**
   * Check if URL is a GitLab URL
   *
   * @param url URL to check
   * @returns True if URL is a GitLab URL
   */
  private isGitLabUrl(url: string): boolean {
    return (
      url.includes('gitlab') ||
      url.includes('-/merge_requests/') ||
      url.includes('/merge_requests/')
    );
  }

  /**
   * Analyze the diff with the configured LLM
   * @param diff - The code diff to analyze
   * @returns The LLM feedback
   */
  async analyzeDiff(diff: string): Promise<string> {
    try {
      const prompt = this.promptTemplate.replace('{{diff}}', diff);
      this.logger.info(`Analyzing diff with ${this.config.llmProvider} LLM...`);

      // Only print the prompt in verbose mode - it can be very large
      if (this.config.verbose) {
        this.logger.debug(`Prompt length: ${prompt.length} characters`);
      }

      return await this.llmAdapter.analyzeCode(prompt);
    } catch (error) {
      this.logger.error('Error analyzing diff with LLM:', error as Error);
      throw error;
    }
  }

  /**
   * Post the review comment on the MR/PR
   * @param feedback - The formatted feedback
   */
  async postComment(feedback: string): Promise<void> {
    try {
      if (this.config.mrMode === 'url') {
        await this.postCommentForUrlMode(feedback);
        return;
      }

      if (this.isRunningInCI()) {
        await this.postCommentForCIMode(feedback);
        return;
      }

      // Default to local mode - just print the feedback
      this.displayLocalModeComment(feedback);
    } catch (error) {
      this.logger.error('Error posting comment:', error as Error);
      throw error;
    }
  }

  /**
   * Post comment for URL mode
   *
   * @param feedback The formatted feedback
   */
  private async postCommentForUrlMode(feedback: string): Promise<void> {
    if (!this.config.gitMrUrl) {
      throw new Error('Git MR/PR URL is required for URL mode');
    }

    const url = this.config.gitMrUrl.toLowerCase();

    if (this.isGitHubUrl(url) && this.config.gitPlatform !== 'github') {
      await this.postCommentWithGitHubAdapter(feedback);
      return;
    }

    if (this.isGitLabUrl(url) && this.config.gitPlatform !== 'gitlab') {
      await this.postCommentWithGitLabAdapter(feedback);
      return;
    }

    // Use the default adapter
    const parsedUrl = this.gitAdapter.parseRequestUrl(this.config.gitMrUrl);
    await this.gitAdapter.commentOnRequest(parsedUrl, feedback);
  }

  /**
   * Post comment using GitHub adapter
   *
   * @param feedback The formatted feedback
   */
  private async postCommentWithGitHubAdapter(feedback: string): Promise<void> {
    this.logger.info('URL appears to be GitHub, using GitHub adapter for commenting.');

    if (!this.config.gitMrUrl) {
      throw new Error('Git MR/PR URL is required for GitHub comment');
    }

    const gitFactory = new GitAdapterFactory();
    const tempConfig = { ...this.config, gitPlatform: 'github' as const };
    const githubAdapter = gitFactory.create(tempConfig);

    const parsedUrl = githubAdapter.parseRequestUrl(this.config.gitMrUrl);
    await githubAdapter.commentOnRequest(parsedUrl, feedback);
  }

  /**
   * Post comment using GitLab adapter
   *
   * @param feedback The formatted feedback
   */
  private async postCommentWithGitLabAdapter(feedback: string): Promise<void> {
    this.logger.info('URL appears to be GitLab, using GitLab adapter for commenting.');

    if (!this.config.gitMrUrl) {
      throw new Error('Git MR/PR URL is required for GitLab comment');
    }

    const gitFactory = new GitAdapterFactory();
    const tempConfig = { ...this.config, gitPlatform: 'gitlab' as const };
    const gitlabAdapter = gitFactory.create(tempConfig);

    const parsedUrl = gitlabAdapter.parseRequestUrl(this.config.gitMrUrl);
    await gitlabAdapter.commentOnRequest(parsedUrl, feedback);
  }

  /**
   * Post comment for CI mode
   *
   * @param feedback The formatted feedback
   */
  private async postCommentForCIMode(feedback: string): Promise<void> {
    if (this.config.gitPlatform === 'gitlab') {
      await this.gitAdapter.commentOnRequest(
        {
          projectId: this.config.projectId || '',
          mergeRequestIid: this.config.mergeRequestIid || '',
        },
        feedback
      );
    } else if (this.config.gitPlatform === 'github') {
      throw new Error('GitHub CI mode not fully implemented yet');
    }
  }

  /**
   * Display comment for local mode
   *
   * @param feedback The formatted feedback
   */
  private displayLocalModeComment(feedback: string): void {
    this.logger.info('Running in local mode, skipping comment creation');
    this.logger.user('\n===== LLM REVIEW ======\n');
    this.logger.user(feedback);
    this.logger.user('\n=======================\n');
  }

  /**
   * Get the LLM model name based on the provider
   */
  private getLLMModelName(): string {
    switch (this.config.llmProvider) {
      case 'anthropic':
        return this.config.anthropicModel;
      case 'openai':
        return this.config.openaiModel;
      case 'deepseek':
        return this.config.deepseekModel;
      case 'copilot':
        return this.config.copilotModel;
      default:
        return 'unknown';
    }
  }
}

File: python/tiw/core/mr_reviewer.py
"""Main class for performing MR/PR reviews."""

import asyncio
import json
import os
from datetime import datetime
from typing import Any, Dict, Optional, Union

from ..adapters.git import GitAdapterFactory
from ..adapters.git.git_adapter import GitAdapter
from ..adapters.llm import LLMAdapterFactory
from ..adapters.llm.llm_adapter import LLMAdapter
from ..config.config import AppConfig, GitPlatform, MRMode
from ..utils.file_utils import FileUtils
from ..utils.logging import Logger
from .review_formatter import ReviewFormatter


class MRReviewer:
    """Main class for performing MR/PR reviews."""
    
    def __init__(self, config: AppConfig):
        """Create a new MR reviewer.
        
        Args:
            config: The loaded and validated configuration
        """
        self.config = config
        self.logger = Logger(config.verbose)
        self.file_utils = FileUtils(self.logger)
        
        # Create adapters using factories
        llm_factory = LLMAdapterFactory()
        git_factory = GitAdapterFactory()
        
        self.llm_adapter: LLMAdapter = llm_factory.create(config)
        self.git_adapter: GitAdapter = git_factory.create(config)
        
        # Create formatter with the configured template
        self.formatter = ReviewFormatter(config.formatter_template)
        
        # Load prompt template from directory
        self.prompt_template = self.file_utils.load_prompt_from_directory(config.prompt_dir)
    
    async def review(self) -> Optional[str]:
        """Execute the review workflow.
        
        Returns:
            Path to the saved review file
        """
        try:
            # Get the diff
            diff = await self.get_diff()
            if self.config.verbose:
                self.logger.debug(f"</diff start>{diff}</diff end>")
            self.logger.info(f"Retrieved diff ({len(diff)} characters)")
            
            if not diff:
                self.logger.warn("No changes detected, skipping LLM analysis")
                return None
            
            if await self.should_cancel_review(diff):
                return None
            
            feedback = await self.analyze_diff(diff)
            self.logger.info("LLM analysis completed")
            
            metadata = self.create_review_metadata()
            
            review_file_path = self.file_utils.save_review_to_file(
                self.config.reviews_dir,
                feedback,
                metadata
            )
            
            parsed_feedback = self.parse_feedback(feedback)
            formatted_comment = self.formatter.format({
                "metadata": metadata,
                "feedback": parsed_feedback
            })
            
            await self.post_comment(formatted_comment)
            
            self.logger.info("Review completed successfully")
            return review_file_path
        
        except Exception as error:
            self.logger.error(f"Review failed: {str(error)}")
            raise
    
    def create_review_metadata(self) -> Dict[str, str]:
        """Create metadata for the review.
        
        Returns:
            Review metadata object
        """
        return {
            "timestamp": datetime.now().isoformat(),
            "llmProvider": self.config.llm_provider,
            "llmModel": self.get_llm_model_name(),
            "mrMode": self.config.mr_mode,
            "gitPlatform": self.config.git_platform,
            "commandLine": " ".join(os.sys.argv)
        }
    
    async def should_cancel_review(self, diff: str) -> bool:
        """Check if the review should be canceled based on user input.
        
        Args:
            diff: The diff content
            
        Returns:
            True if the review should be canceled
        """
        if not self.config.show_diff:
            return False
        
        self.display_diff(diff)
        
        # If we're not in interactive mode, continue with the review
        if not os.isatty(0):  # 0 is stdin file descriptor
            return False
        
        should_cancel = await self.prompt_for_cancellation()
        if should_cancel:
            self.logger.info("LLM analysis cancelled by user")
            return True
        
        return False
    
    def display_diff(self, diff: str):
        """Display the diff to the user.
        
        Args:
            diff: The diff content
        """
        self.logger.user("\n===== DIFF START =====\n")
        self.logger.user(diff)
        self.logger.user("\n===== DIFF END =====\n")
    
    async def prompt_for_cancellation(self) -> bool:
        """Prompt the user for cancellation.
        
        Returns:
            True if the user wants to cancel
        """
        import sys
        
        self.logger.user("Continue with LLM analysis? (Y/n): ")
        sys.stdout.flush()  # Ensure prompt is displayed
        
        answer = input()
        return answer.lower() == 'n'
    
    def parse_feedback(self, feedback: Union[str, Dict]) -> Dict[str, Any]:
        """Parse the feedback string into a structured object.
        
        Args:
            feedback: The feedback string or object from the LLM
            
        Returns:
            Parsed feedback object
        """
        try:
            parsed_feedback = feedback if isinstance(feedback, dict) else json.loads(feedback)
            
            # Create a new object with default values and merge with parsed feedback
            result = {
                # Default values
                "overview": parsed_feedback.get("overview", {}),
                "fileReviews": parsed_feedback.get("fileReviews", []),
                "testReview": parsed_feedback.get("testReview", ""),
                "generalFeedback": parsed_feedback.get("generalFeedback", ""),
            }
            
            # Copy any other properties
            for key, value in parsed_feedback.items():
                if key not in ["overview", "fileReviews", "testReview", "generalFeedback"]:
                    result[key] = value
            
            return result
        
        except Exception as error:
            self.logger.error(f"Error parsing feedback for formatting: {str(error)}")
            raise ValueError("Could not format review due to JSON parsing error")
    
    async def get_diff(self) -> str:
        """Get MR/PR diff based on the configured mode.
        
        Returns:
            The diff content
        """
        try:
            if self.is_running_in_ci():
                return await self.get_ci_diff()
            
            if self.config.mr_mode == MRMode.URL:
                return await self.get_url_diff()
            
            # Default to local mode
            self.logger.info("Running in local mode, fetching diff from local git")
            return await self.git_adapter.get_local_diff()
        
        except Exception as error:
            self.logger.error(f"Error getting MR/PR diff: {str(error)}")
            raise
    
    def is_running_in_ci(self) -> bool:
        """Check if running in CI mode.
        
        Returns:
            True if running in CI mode
        """
        return (
            self.config.mr_mode == MRMode.CI and 
            os.environ.get('CI_PIPELINE_SOURCE') == 'merge_request_event'
        )
    
    async def get_ci_diff(self) -> str:
        """Get diff for CI mode.
        
        Returns:
            The diff content
        """
        self.logger.info("Running in CI mode, fetching diff from Git platform API")
        
        if self.config.git_platform == GitPlatform.GITLAB:
            return await self.git_adapter.get_request_diff({
                "projectId": self.config.project_id or "",
                "mergeRequestIid": self.config.merge_request_iid or "",
            })
        
        if self.config.git_platform == GitPlatform.GITHUB:
            raise ValueError("GitHub CI mode not fully implemented yet")
        
        raise ValueError("Unsupported Git platform for CI mode")
    
    async def get_url_diff(self) -> str:
        """Get diff for URL mode.
        
        Returns:
            The diff content
        """
        if not self.config.git_mr_url:
            raise ValueError("Git MR/PR URL is required for URL mode")
        
        url = self.config.git_mr_url.lower()
        actual_platform = self.config.git_platform
        
        if self.is_github_url(url):
            if self.config.git_platform != GitPlatform.GITHUB:
                return await self.get_github_diff_with_temp_adapter()
        
        if self.is_gitlab_url(url):
            if self.config.git_platform != GitPlatform.GITLAB:
                return await self.get_gitlab_diff_with_temp_adapter()
        
        self.logger.info(f"Running in URL mode, fetching diff from {actual_platform} API using URL")
        
        # Use the default adapter
        parsed_url = self.git_adapter.parse_request_url(self.config.git_mr_url)
        return await self.git_adapter.get_request_diff(parsed_url)
    
    async def get_github_diff_with_temp_adapter(self) -> str:
        """Get GitHub diff using a temporary adapter.
        
        Returns:
            The diff content
        """
        self.logger.info(
            "URL appears to be GitHub, but platform is set to GitLab. Switching to GitHub adapter."
        )
        
        if not self.config.git_mr_url:
            raise ValueError("Git MR/PR URL is required for GitHub adapter")
        
        git_factory = GitAdapterFactory()
        temp_config = AppConfig(
            **vars(self.config),
            git_platform=GitPlatform.GITHUB
        )
        github_adapter = git_factory.create(temp_config)
        
        parsed_url = github_adapter.parse_request_url(self.config.git_mr_url)
        return await github_adapter.get_request_diff(parsed_url)
    
    async def get_gitlab_diff_with_temp_adapter(self) -> str:
        """Get GitLab diff using a temporary adapter.
        
        Returns:
            The diff content
        """
        self.logger.info(
            "URL appears to be GitLab, but platform is set to GitHub. Switching to GitLab adapter."
        )
        
        if not self.config.git_mr_url:
            raise ValueError("Git MR/PR URL is required for GitLab adapter")
        
        git_factory = GitAdapterFactory()
        temp_config = AppConfig(
            **vars(self.config),
            git_platform=GitPlatform.GITLAB
        )
        gitlab_adapter = git_factory.create(temp_config)
        
        parsed_url = gitlab_adapter.parse_request_url(self.config.git_mr_url)
        diff_result = await gitlab_adapter.get_request_diff(parsed_url)
        
        if self.config.verbose:
            self.logger.debug(f"</diff start>{diff_result}</diff end>")
        
        return diff_result
    
    def is_github_url(self, url: str) -> bool:
        """Check if URL is a GitHub URL.
        
        Args:
            url: URL to check
            
        Returns:
            True if URL is a GitHub URL
        """
        return 'github.com' in url or '/pull/' in url
    
    def is_gitlab_url(self, url: str) -> bool:
        """Check if URL is a GitLab URL.
        
        Args:
            url: URL to check
            
        Returns:
            True if URL is a GitLab URL
        """
        return (
            'gitlab' in url or
            '-/merge_requests/' in url or
            '/merge_requests/' in url
        )
    
    async def analyze_diff(self, diff: str) -> str:
        """Analyze the diff with the configured LLM.
        
        Args:
            diff: The code diff to analyze
            
        Returns:
            The LLM feedback
        """
        try:
            prompt = self.prompt_template.replace('{{diff}}', diff)
            self.logger.info(f"Analyzing diff with {self.config.llm_provider} LLM...")
            
            # Only print the prompt in verbose mode - it can be very large
            if self.config.verbose:
                self.logger.debug(f"Prompt length: {len(prompt)} characters")
            
            return await self.llm_adapter.analyze_code(prompt)
        
        except Exception as error:
            self.logger.error(f"Error analyzing diff with LLM: {str(error)}")
            raise
    
    async def post_comment(self, feedback: str) -> None:
        """Post the review comment on the MR/PR.
        
        Args:
            feedback: The formatted feedback
        """
        try:
            if self.config.mr_mode == MRMode.URL:
                await self.post_comment_for_url_mode(feedback)
                return
            
            if self.is_running_in_ci():
                await self.post_comment_for_ci_mode(feedback)
                return
            
            # Default to local mode - just print the feedback
            self.display_local_mode_comment(feedback)
        
        except Exception as error:
            self.logger.error(f"Error posting comment: {str(error)}")
            raise
    
    async def post_comment_for_url_mode(self, feedback: str) -> None:
        """Post comment for URL mode.
        
        Args:
            feedback: The formatted feedback
        """
        if not self.config.git_mr_url:
            raise ValueError("Git MR/PR URL is required for URL mode")
        
        url = self.config.git_mr_url.lower()
        
        if self.is_github_url(url) and self.config.git_platform != GitPlatform.GITHUB:
            await self.post_comment_with_github_adapter(feedback)
            return
        
        if self.is_gitlab_url(url) and self.config.git_platform != GitPlatform.GITLAB:
            await self.post_comment_with_gitlab_adapter(feedback)
            return
        
        # Use the default adapter
        parsed_url = self.git_adapter.parse_request_url(self.config.git_mr_url)
        await self.git_adapter.comment_on_request(parsed_url, feedback)
    
    async def post_comment_with_github_adapter(self, feedback: str) -> None:
        """Post comment using GitHub adapter.
        
        Args:
            feedback: The formatted feedback
        """
        self.logger.info("URL appears to be GitHub, using GitHub adapter for commenting.")
        
        if not self.config.git_mr_url:
            raise ValueError("Git MR/PR URL is required for GitHub comment")
        
        git_factory = GitAdapterFactory()
        temp_config = AppConfig(
            **vars(self.config),
            git_platform=GitPlatform.GITHUB
        )
        github_adapter = git_factory.create(temp_config)
        
        parsed_url = github_adapter.parse_request_url(self.config.git_mr_url)
        await github_adapter.comment_on_request(parsed_url, feedback)
    
    async def post_comment_with_gitlab_adapter(self, feedback: str) -> None:
        """Post comment using GitLab adapter.
        
        Args:
            feedback: The formatted feedback
        """
        self.logger.info("URL appears to be GitLab, using GitLab adapter for commenting.")
        
        if not self.config.git_mr_url:
            raise ValueError("Git MR/PR URL is required for GitLab comment")
        
        git_factory = GitAdapterFactory()
        temp_config = AppConfig(
            **vars(self.config),
            git_platform=GitPlatform.GITLAB
        )
        gitlab_adapter = git_factory.create(temp_config)
        
        parsed_url = gitlab_adapter.parse_request_url(self.config.git_mr_url)
        await gitlab_adapter.comment_on_request(parsed_url, feedback)
    
    async def post_comment_for_ci_mode(self, feedback: str) -> None:
        """Post comment for CI mode.
        
        Args:
            feedback: The formatted feedback
        """
        if self.config.git_platform == GitPlatform.GITLAB:
            await self.git_adapter.comment_on_request(
                {
                    "projectId": self.config.project_id or "",
                    "mergeRequestIid": self.config.merge_request_iid or "",
                },
                feedback
            )
        elif self.config.git_platform == GitPlatform.GITHUB:
            raise ValueError("GitHub CI mode not fully implemented yet")
    
    def display_local_mode_comment(self, feedback: str) -> None:
        """Display comment for local mode.
        
        Args:
            feedback: The formatted feedback
        """
        self.logger.info("Running in local mode, skipping comment creation")
        self.logger.user("\n===== LLM REVIEW ======\n")
        self.logger.user(feedback)
        self.logger.user("\n=======================\n")
    
    def get_llm_model_name(self) -> str:
        """Get the LLM model name based on the provider.
        
        Returns:
            The model name
        """
        if self.config.llm_provider == "anthropic":
            return self.config.anthropic_model
        elif self.config.llm_provider == "openai":
            return self.config.openai_model
        elif self.config.llm_provider == "deepseek":
            return self.config.deepseek_model
        elif self.config.llm_provider == "copilot":
            return self.config.copilot_model
        else:
            return "unknown"